File: lib/py/dsu.py
class DSU(object):
    def __init__(self, n):
        self.ids = range(n)
        self.sizes = [1] * n

    def root(self, x):
        while x != self.ids[x]:
            x = self.ids[x]
        return x

    def is_connected(self, u, v):
        return self.root(u) == self.root(v)

    def connect(self, u, v):
        ru = self.root(u)
        rv = self.root(v)
        if self.sizes[ru] < self.sizes[rv]:
            self.ids[ru] = rv
            self.sizes[rv] += self.sizes[ru]
        else:
            self.ids[rv] = ru
            self.sizes[ru] += self.sizes[rv]

File: lib/cpp/dsu.cpp
#include <vector>


struct DSU {
    vector<int> ids;
    vector<int> sizes;

    DSU(int n) {
        ids.resize(n);
        for (int i = 0; i < n; ++i)
            ids[i] = i;
        sizes.resize(n, 1);
    }

    int root(int x) {
        while (x != ids[x])
            x = ids[x];
        return x;
    }

    int is_connected(int u, int v) {
        return root(u) == root(v);
    }

    void connect(int u, int v) {
        int ru = root(u);
        int rv = root(v);
        if (ru == rv) {
            return;
        }
        if (sizes[ru] < sizes[rv]) {
            ids[ru] = rv;
            sizes[rv] += sizes[ru];
        }
        else {
            ids[rv] = ru;
            sizes[ru] += sizes[rv];
        }
    }
};
